mtt/src/mtt.rs

```diff
@@ -0,0 +1,1190 @@
+// and '01b as confidential ('11b is illegal).
+struct MttL1Page<'a> {
+    base: [u64; L1_ENTRY_COUNT],
+    marker: PhantomData<&'a u64>,
+}
+
+impl<'a> MttL1Page<'a> {
+    // Returns a u64 slice that allows access L1 page entries represented by pfn.
+    // #Safety:
+    // The caller must guarantee that the pfn points to a valid L1 page that isn't
+    // aliased elsewhere, and points to confidential memory. The page must remain
+    // allocated for the lifetime of the entry in the L2 table.
+    unsafe fn l1_slice_from_pfn(pfn: SupervisorPfn) -> &'a mut [u64] {
+        let l1_page_addr = (pfn.bits() << 12) as *mut MttL1Page;
+        let l1_page = &mut *core::ptr::slice_from_raw_parts_mut(l1_page_addr, 1);
+        &mut l1_page[0].base[0..]
+    }
+}
+
+/// Holds references to the platform allocated L2 and L2 tables in a Mutex.
+/// This should be constructed by calling init() with the platform allocated backing
+/// memory for the L2 and L1 tables.
+pub struct Mtt {
+    inner: Mutex<MttL2Directory>,
+}
+
+// Singleton instance of Mtt constructed by Mtt::init()
+static PLATFORM_MTT_TABLES: Once<Mtt> = Once::new();
+
+/// Exposes functions to change and query the type of memory ranges in the MTT.
+/// Successful invocation requires a preceding call to Mtt::init().
+impl Mtt {
+    // Performs sanity checks on the platform L2 table
+    #[cfg(not(test))]
+    fn validate_mtt(&self) -> Result<()> {
+        let l2_dir = self.inner.lock();
+        let l2_base_addr = l2_dir.mtt_l2_base.as_ptr() as usize;
+        let addr = SupervisorPageAddr::new(RawAddr::supervisor(l2_base_addr as u64))
+            .ok_or(InvalidAddress)?;
+        // TODO: Add additional checks to validate individual L2 entries
+        let mut mtt_range = MttRange::new(addr, L2_BYTE_SIZE, l2_dir)?;
+        while let Some((addr, region_size)) = mtt_range.next() {
+            let region_memory_type = mtt_range.get_region_memory_type(addr, region_size)?;
+            if region_memory_type != MttMemoryType::Confidential {
+                return Err(InvalidPlatformConfiguration);
+            }
+        }
+        Ok(())
+    }
+
+    // Test only NOP validation
+    #[cfg(test)]
+    fn validate_mtt(&self) -> Result<()> {
+        let l2_dir = self.inner.lock();
+        // Intended only to silence clippy warning about unused imports while testing
+        let l2_base_addr = l2_dir.mtt_l2_base.as_ptr() as u64;
+        SupervisorPageAddr::new(RawAddr::supervisor(l2_base_addr)).ok_or(InvalidAddress)?;
+        Ok(())
+    }
+
+    /// Creates the singleton Mtt instance.
+    /// # Safety
+    ///
+    /// The caller must ensure that:
+    /// `mtt_l2_base` is a platform allocated 8MB memory region that isn't
+    /// aliased elsewhere and must remain allocated until the platform resets.
+    /// The function performs some basic sanity checks on the passed-in parameters.
+    pub unsafe fn init(l2_base_addr: u64, len: usize) -> Result<()> {
```

Comment:
i should've mentioned this in the last round but i totally forgot: this can be made "safe" if you use a `SequentialPages` since `SequentialPages` is an owning reference to a contiguous slice of pages.